### PR TITLE
[FIX] Prevent duplicate cyborg upgrades

### PR DESCRIPTION
**Pull request by Adrer**
## What Does This PR Do
Prevents you from installing duplicate cyborg upgrades.

Fixes: #26715
Fixes: #26557

## Why It's Good For The Game
You should not be able to install multiple of the same upgrade.

## Testing
Attempt to install 2 VTEC upgrades, the second one fails.
Attempt to install 2 kinetic accelerator damage modules, works just fine.

<hr>

### Declaration

- [x] I confirm that I either do not require [pre-approval](https://github.com/ParadiseSS13/Paradise/blob/master/docs/CODE_OF_CONDUCT.md#types-of-changes-that-need-approval) for this PR, or I have obtained such approval and have included a screenshot to demonstrate this below.
  
  
  <hr>

## Changelog

:cl:
fix: You can no longer install duplicate upgrades into cyborgs
/:cl:

### CODE REVIEW

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 1]
```diff
@@ -17,19 +17,30 @@
 	var/list/items_to_add = list()
 	/// A list of replacement items will need to be placed into a cyborg module's `special_rechargable` list after this upgrade is installed.
 	var/list/special_rechargables = list()
+	/// Allow the same upgrade to be installed multiple times, FALSE by default
+	var/allow_duplicate = FALSE
+	/// Delete the module after installing it. For deleting upgrades that might be install multiple times like the rename/reset upgrades.
```
Comment (Fordoxia) [inline code/game/objects/items/robot/robot_upgrades.dm:22]:
```suggestion
	/// Delete the module after installing it. For deleting upgrades that might be installed multiple times, like the rename/reset upgrades.
```

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 2]
```diff
@@ -43,9 +54,13 @@
 		to_chat(usr, "<span class='warning'>[src] will not function on a deceased cyborg.</span>")
 		return
 	if(module_type && !istype(R.module, module_type))
-		to_chat(R, "<span class='warning'>Upgrade mounting error!  No suitable hardpoint detected!</span>")
+		to_chat(R, "<span class='warning'>Upgrade mounting error! No suitable hardpoint detected!</span>")
 		to_chat(usr, "<span class='warning'>There's no mounting point for the module!</span>")
 		return
+	var/obj/item/borg/upgrade/u = locate(type) in R
+	if(u && !allow_duplicate)
+		to_chat(usr, "<span class='notice'>This unit already has [src] installed!</span>")
```
Comment (Fordoxia) [inline code/game/objects/items/robot/robot_upgrades.dm:62]: I'm pretty sure that `mob/user` is preferred over `usr`.

Comment (DGamerL) [reply code/game/objects/items/robot/robot_upgrades.dm:62]: Less preferred and more required

**Review by BiancaWilkson — COMMENTED** [review]
Make sure after_install happens after all the other checks.

**Comment by Truls20** [comment]
This PR will also fix #26557

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 3]
```diff
@@ -17,19 +17,30 @@
...
  */
-/obj/item/borg/upgrade/proc/action(mob/living/silicon/robot/R)
+/obj/item/borg/upgrade/proc/action(mob/user, mob/living/silicon/robot/R)
 	if(!pre_install_checks(R))
 		return
 	if(!do_install(R))
 		return
 	after_install(R)
+	if(!user.drop_item())
+		to_chat(user, "<span class='notice'>[src] is stuck to your hand, you cannot install it in [R]</span>")
```
Comment (lewcc) [inline code/game/objects/items/robot/robot_upgrades.dm:38]:
```suggestion
		to_chat(user, "<span class='notice'>[src] is stuck to your hand, you cannot install it in [R].</span>")
```

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 4]
```diff
@@ -17,19 +17,30 @@
...
+		return
+	if(delete_after_install)
+		qdel(src)
+	else
+		forceMove(R)
```
[the remark this answers is not in the document]
Comment (lewcc) [reply code/game/objects/items/robot/robot_upgrades.dm:43]: isn't after_install already called earlier?

[the remark this answers is not in the document]
Comment (BiancaWilkson) [reply code/game/objects/items/robot/robot_upgrades.dm:43]: For some reason github split these reviews into 2, but the review above this one is deleting after_install so it moves down here.

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 5]
```diff
@@ -17,19 +17,30 @@
...
  * Arguments:
  * * R - the cyborg that was clicked on with an upgrade.
  */
-/obj/item/borg/upgrade/proc/action(mob/living/silicon/robot/R)
+/obj/item/borg/upgrade/proc/action(mob/user, mob/living/silicon/robot/R)
 	if(!pre_install_checks(R))
 		return
 	if(!do_install(R))
 		return
 	after_install(R)
```
[the remark this answers is not in the document]
Comment (Adrer) [reply code/game/objects/items/robot/robot_upgrades.dm:36]: The after_install should be performed before the module is (potentially) deleted though

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 6]
```diff
@@ -17,19 +17,30 @@
...
+	if(!user.drop_item())
+		to_chat(user, "<span class='notice'>[src] is stuck to your hand, you cannot install it in [R].</span>")
+		return
+	if(delete_after_install)
```
Comment (BiancaWilkson) [inline code/game/objects/items/robot/robot_upgrades.dm:40]:
```suggestion
		return
	if(!user.drop_item())
		to_chat(user, "<span class='notice'>[src] is stuck to your hand, you cannot install it in [R].</span>")
		return
    after_install(R)
	if(delete_after_install)
```

File: code/game/objects/items/robot/robot_upgrades.dm (hunk 5 above)
[the remark this answers is not in the document]
Comment (BiancaWilkson) [reply code/game/objects/items/robot/robot_upgrades.dm:36]: Good point, I'll revise the suggestion.

**Review by BiancaWilkson — CHANGES_REQUESTED** [review]

**Review by DGamerL — APPROVED** [review]

**Pull request merged**